File: ScaleBench/Protocols/WMBPlusUSBSerialParser.swift
import Foundation

struct WMBPlusUSBStatus: OptionSet, Equatable {
    let rawValue: UInt16

    static let hx711Connected = WMBPlusUSBStatus(rawValue: 0x0001)
    static let bleConnected = WMBPlusUSBStatus(rawValue: 0x0002)
    static let recentBump = WMBPlusUSBStatus(rawValue: 0x0004)
    static let recentGlitch = WMBPlusUSBStatus(rawValue: 0x0008)
    static let zeroClamped = WMBPlusUSBStatus(rawValue: 0x0010)
    static let autoZeroActive = WMBPlusUSBStatus(rawValue: 0x0020)
    static let batteryValid = WMBPlusUSBStatus(rawValue: 0x0040)
    static let charging = WMBPlusUSBStatus(rawValue: 0x0080)
    static let wifiRadioOn = WMBPlusUSBStatus(rawValue: 0x0100)

    var labels: [String] {
        [
            (Self.hx711Connected, "HX711 connected"),
            (Self.bleConnected, "BLE connected"),
            (Self.recentBump, "Recent bump"),
            (Self.recentGlitch, "Recent glitch"),
            (Self.zeroClamped, "Zero clamped"),
            (Self.autoZeroActive, "Auto-zero active"),
            (Self.batteryValid, "Battery valid"),
            (Self.charging, "Charging"),
            (Self.wifiRadioOn, "WiFi radio on"),
        ].compactMap { flag, label in contains(flag) ? label : nil }
    }
}

enum WMBPlusUSBSerialParseError: Equatable {
    case fieldCount
    case invalidInteger(field: String)
    case invalidFloat(field: String)
    case invalidStatus
    case invalidRange(field: String)
}

enum WMBPlusUSBSerialParseResult: Equatable {
    case ignored
    case rejected(WMBPlusUSBSerialParseError)
    case sample(WMBPlusUSBSerialRow)
}

struct WMBPlusUSBSerialRow: Equatable {
    static let protocolName = "WMB+ USB Serial"
    static let deviceName = "WMB+ USB"
    static let baud = 115_200

    var firmwareMillis: UInt32
    var sequenceNumber: UInt32
    var weightGrams: Double
    var flowGramsPerSecond: Double
    var status: WMBPlusUSBStatus
    var firmwareQuality: Int
    var batteryPercent: Int?
    var hx711Hz: Double
    var droppedCumulative: UInt32
    var droppedDelta: UInt32
    var hostReceivedAt: Date
    var hostMonotonicSeconds: Double
    var fields: [PacketFieldAnnotation]

    var isValidWeightSample: Bool {
        status.contains(.hx711Connected)
    }

    var metadata: USBSerialSampleMetadata {
        USBSerialSampleMetadata(
            firmwareMillis: firmwareMillis,
            sequenceNumber: sequenceNumber,
            usbStatusRaw: status.rawValue,
            usbStatusLabels: status.labels,
            firmwareQuality: firmwareQuality,
            hx711Hz: hx711Hz,
            usbDroppedCumulative: droppedCumulative,
            usbDroppedDelta: droppedDelta,
            hostReceivedAt: hostReceivedAt
        )
    }

    var sample: ScaleSample {
        var statusByte: UInt8 = 0
        if status.contains(.hx711Connected) { statusByte |= 0x02 }
        if status.contains(.batteryValid) { statusByte |= 0x40 }

        var diagnosticByte: UInt8 = 0x20 | 0x40 | 0x80
        if status.contains(.recentBump) { diagnosticByte |= 0x01 }
        if hx711Hz > 0 { diagnosticByte |= 0x04 }
        if hx711Hz >= 60 { diagnosticByte |= 0x08 }
        if (8...12).contains(hx711Hz) { diagnosticByte |= 0x10 }

        return ScaleSample(
            arrivalTime: hostReceivedAt,
            monotonicSeconds: hostMonotonicSeconds,
            scaleKind: .weighMyBruPlus,
            weightGrams: weightGrams,
            deviceTimestampMilliseconds: firmwareMillis,
            sequence: nil,
            batteryPercent: batteryPercent,
            flowGramsPerSecond: flowGramsPerSecond,
            firmwareQualityScore: firmwareQuality,
            detectedSampleRateHz: Int(hx711Hz.rounded()),
            statusFlags: ScaleStatusFlags(byte: statusByte),
            diagnosticFlags: ScaleDiagnosticFlags(byte: diagnosticByte),
            usbSerial: metadata
        )
    }
}

struct WMBPlusUSBSerialParser {
    static let header = "WMBP_WEIGHT_V1_HEADER,ms,seq,weight_g,flow_gps,status,quality,battery_pct,hx711_hz,dropped"
    private(set) var previousDropped: UInt32?

    mutating func reset() {
        previousDropped = nil
    }

    mutating func parse(
        line: String,
        hostReceivedAt: Date,
        hostMonotonicSeconds: Double
    ) -> WMBPlusUSBSerialParseResult {
        guard line.hasPrefix("WMBP_WEIGHT_V1,") else { return .ignored }

        let values = line.split(separator: ",", omittingEmptySubsequences: false).map(String.init)
        guard values.count == 10 else { return .rejected(.fieldCount) }

        guard let firmwareMillis = UInt32(values[1]) else {
            return .rejected(.invalidInteger(field: "ms"))
        }
        guard let sequenceNumber = UInt32(values[2]) else {
            return .rejected(.invalidInteger(field: "seq"))
        }
        guard let weightGrams = finiteDouble(values[3]) else {
            return .rejected(.invalidFloat(field: "weight_g"))
        }
        guard let flowGramsPerSecond = finiteDouble(values[4]) else {
            return .rejected(.invalidFloat(field: "flow_gps"))
        }
        guard values[5].hasPrefix("0x") || values[5].hasPrefix("0X"),
              let statusRaw = UInt16(values[5].dropFirst(2), radix: 16) else {
            return .rejected(.invalidStatus)
        }
        let status = WMBPlusUSBStatus(rawValue: statusRaw)
        guard let firmwareQuality = Int(values[6]) else {
            return .rejected(.invalidInteger(field: "quality"))
        }
        guard (0...100).contains(firmwareQuality) else {
            return .rejected(.invalidRange(field: "quality"))
        }
        guard let rawBattery = Int(values[7]) else {
            return .rejected(.invalidInteger(field: "battery_pct"))
        }
        if status.contains(.batteryValid), !(0...100).contains(rawBattery) {
            return .rejected(.invalidRange(field: "battery_pct"))
        }
        guard let hx711Hz = finiteDouble(values[8]) else {
            return .rejected(.invalidFloat(field: "hx711_hz"))
        }
        guard hx711Hz >= 0 else {
            return .rejected(.invalidRange(field: "hx711_hz"))
        }
        guard let droppedCumulative = UInt32(values[9]) else {
            return .rejected(.invalidInteger(field: "dropped"))
        }

        let droppedDelta: UInt32
        if let previousDropped, droppedCumulative > previousDropped {
            droppedDelta = droppedCumulative - previousDropped
        } else {
            droppedDelta = 0
        }
        previousDropped = droppedCumulative

        return .sample(WMBPlusUSBSerialRow(
            firmwareMillis: firmwareMillis,
            sequenceNumber: sequenceNumber,
            weightGrams: weightGrams,
            flowGramsPerSecond: flowGramsPerSecond,
            status: status,
            firmwareQuality: firmwareQuality,
            batteryPercent: status.contains(.batteryValid) ? rawBattery : nil,
            hx711Hz: hx711Hz,
            droppedCumulative: droppedCumulative,
            droppedDelta: droppedDelta,
            hostReceivedAt: hostReceivedAt,
            hostMonotonicSeconds: hostMonotonicSeconds,
            fields: annotations(for: values)
        ))
    }

    private func finiteDouble(_ value: String) -> Double? {
        guard let parsed = Double(value), parsed.isFinite else { return nil }
        return parsed
    }

    private func annotations(for values: [String]) -> [PacketFieldAnnotation] {
        let labels: [(String, PacketFieldSemantic)] = [
            ("Type", .header),
            ("Firmware millis", .timestamp),
            ("Sequence", .sequence),
            ("Weight", .weight),
            ("Flow", .flow),
            ("Status", .status),
            ("Firmware quality", .quality),
            ("Battery", .battery),
            ("HX711 cadence", .sampleRate),
            ("USB dropped", .payload),
        ]
        var byteOffset = 0
        return zip(values, labels).map { value, descriptor in
            defer { byteOffset += value.utf8.count + 1 }
            return PacketFieldAnnotation(
                startByte: byteOffset,
                endByteExclusive: byteOffset + value.utf8.count,
                label: descriptor.0,
                decodedValue: value,
                semantic: descriptor.1
            )
        }
    }
}

struct USBSerialLineBuffer {
    private var bytes = Data()
    private let maximumBufferedBytes = 64 * 1024

    mutating func append(_ chunk: Data) -> [String] {
        bytes.append(chunk)
        if bytes.count > maximumBufferedBytes,
           !bytes.contains(0x0A) {
            bytes.removeAll(keepingCapacity: true)
            return []
        }

        var lines: [String] = []
        while let newline = bytes.firstIndex(of: 0x0A) {
            var lineData = bytes[..<newline]
            if lineData.last == 0x0D { lineData = lineData.dropLast() }
            bytes.removeSubrange(...newline)
            if let line = String(data: lineData, encoding: .utf8) {
                lines.append(line)
            }
        }
        return lines
    }

    mutating func reset() {
        bytes.removeAll(keepingCapacity: true)
    }
}

File: ScaleBenchAndroid/app/src/main/java/app/scalebench/android/WMBPlusUSBSerialParser.java
package app.scalebench.android;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

final class WMBPlusUSBSerialParser {
	    static final String PROTOCOL_NAME = "WMB+ USB Serial";
	    static final String DEVICE_NAME = "WMB+ USB";
	    static final int BAUD = 115_200;
	    static final String HEADER = "WMBP_WEIGHT_V1_HEADER,ms,seq,weight_g,flow_gps,status,quality,battery_pct,hx711_hz,dropped";
	    static final int STATUS_HX711_CONNECTED = 0x0001;

    private Long previousDropped;

    void reset() {
        previousDropped = null;
    }

    ParseResult parse(String line, long hostReceivedAtMillis, double hostMonotonicSeconds) {
        if (!line.startsWith("WMBP_WEIGHT_V1,")) {
            return ParseResult.ignored();
        }
        String[] values = line.split(",", -1);
        if (values.length != 10) {
            return ParseResult.rejected("Expected 10 CSV fields");
        }

        try {
            long firmwareMillis = parseUnsigned32(values[1], "ms");
            long sequenceNumber = parseUnsigned32(values[2], "seq");
            double weightGrams = finiteDouble(values[3], "weight_g");
            double flowGramsPerSecond = finiteDouble(values[4], "flow_gps");
            int statusRaw = parseStatus(values[5]);
            int firmwareQuality = parseIntRange(values[6], "quality", 0, 100);
            int rawBattery = parseInt(values[7], "battery_pct");
            boolean batteryValid = has(statusRaw, 0x0040);
            if (batteryValid && (rawBattery < 0 || rawBattery > 100)) {
                return ParseResult.rejected("battery_pct outside 0...100");
            }
            double hx711Hz = finiteDouble(values[8], "hx711_hz");
            if (hx711Hz < 0) {
                return ParseResult.rejected("hx711_hz must be nonnegative");
            }
            long droppedCumulative = parseUnsigned32(values[9], "dropped");
            long droppedDelta = 0;
            if (previousDropped != null && droppedCumulative > previousDropped) {
                droppedDelta = droppedCumulative - previousDropped;
            }
            previousDropped = droppedCumulative;

            USBSerialSampleMetadata metadata = new USBSerialSampleMetadata();
            metadata.firmwareMillis = firmwareMillis;
            metadata.sequenceNumber = sequenceNumber;
            metadata.usbStatusRaw = statusRaw;
            metadata.usbStatusLabels.addAll(statusLabels(statusRaw));
            metadata.firmwareQuality = firmwareQuality;
            metadata.hx711Hz = hx711Hz;
            metadata.usbDroppedCumulative = droppedCumulative;
            metadata.usbDroppedDelta = droppedDelta;
            metadata.hostReceivedAtMillis = hostReceivedAtMillis;

            ScaleSample sample = new ScaleSample();
            sample.arrivalTimeMillis = hostReceivedAtMillis;
            sample.monotonicSeconds = hostMonotonicSeconds;
            sample.scaleKind = ScaleKind.WEIGH_MY_BRU_PLUS;
            sample.weightGrams = weightGrams;
            sample.deviceTimestampMilliseconds = firmwareMillis;
            sample.sequence = null;
            sample.batteryPercent = batteryValid ? rawBattery : null;
            sample.flowGramsPerSecond = flowGramsPerSecond;
            sample.firmwareQualityScore = firmwareQuality;
            sample.detectedSampleRateHz = hx711Hz > 0 ? (int) Math.round(hx711Hz) : null;
            sample.statusFlags = new ScaleStatusFlags(statusFlagsByte(statusRaw));
            sample.diagnosticFlags = new ScaleDiagnosticFlags(diagnosticFlagsByte(statusRaw, hx711Hz));
            sample.usbSerial = metadata;

            RawScalePacket packet = new RawScalePacket();
            packet.arrivalTimeMillis = hostReceivedAtMillis;
            packet.monotonicSeconds = hostMonotonicSeconds;
            packet.scaleKind = ScaleKind.WEIGH_MY_BRU_PLUS;
            packet.characteristicUuid = PROTOCOL_NAME;
            packet.role = PacketRole.WEIGHT;
            packet.bytesHex = ScaleParsers.hex(line.getBytes(StandardCharsets.UTF_8));
            packet.weightGrams = weightGrams;
            packet.sequence = null;
            packet.deviceTimestampMilliseconds = firmwareMillis;
            packet.usbSerial = metadata;
            packet.fields.addAll(annotations(values));

            return ParseResult.sample(sample, packet);
        } catch (IllegalArgumentException error) {
            return ParseResult.rejected(error.getMessage());
        }
    }

    private static int statusFlagsByte(int statusRaw) {
        int value = 0;
        if (has(statusRaw, 0x0001)) value |= 0x02;
        if (has(statusRaw, 0x0040)) value |= 0x40;
        return value;
    }

    private static int diagnosticFlagsByte(int statusRaw, double hx711Hz) {
        int value = 0x20 | 0x40 | 0x80;
        if (has(statusRaw, 0x0004)) value |= 0x01;
        if (hx711Hz > 0) value |= 0x04;
        if (hx711Hz >= 60) value |= 0x08;
        if (hx711Hz >= 8 && hx711Hz <= 12) value |= 0x10;
        return value;
    }

    private static List<String> statusLabels(int statusRaw) {
        List<String> labels = new ArrayList<>();
        addLabel(labels, statusRaw, 0x0001, "HX711 connected");
        addLabel(labels, statusRaw, 0x0002, "BLE connected");
        addLabel(labels, statusRaw, 0x0004, "Recent bump");
        addLabel(labels, statusRaw, 0x0008, "Recent glitch");
        addLabel(labels, statusRaw, 0x0010, "Zero clamped");
        addLabel(labels, statusRaw, 0x0020, "Auto-zero active");
        addLabel(labels, statusRaw, 0x0040, "Battery valid");
        addLabel(labels, statusRaw, 0x0080, "Charging");
        addLabel(labels, statusRaw, 0x0100, "WiFi radio on");
        return labels;
    }

    private static void addLabel(List<String> labels, int statusRaw, int flag, String label) {
        if (has(statusRaw, flag)) labels.add(label);
    }

    private static boolean has(int value, int flag) {
        return (value & flag) != 0;
    }

    private static long parseUnsigned32(String value, String field) {
        try {
            long parsed = Long.parseLong(value);
            if (parsed < 0 || parsed > 0xFFFF_FFFFL) {
                throw new IllegalArgumentException(field + " outside UInt32 range");
            }
            return parsed;
        } catch (NumberFormatException error) {
            throw new IllegalArgumentException(field + " is not an integer");
        }
    }

    private static int parseStatus(String value) {
        if (!value.startsWith("0x") && !value.startsWith("0X")) {
            throw new IllegalArgumentException("status must be hex");
        }
        try {
            int parsed = Integer.parseInt(value.substring(2), 16);
            if (parsed < 0 || parsed > 0xFFFF) {
                throw new IllegalArgumentException("status outside UInt16 range");
            }
            return parsed;
        } catch (NumberFormatException error) {
            throw new IllegalArgumentException("status must be hex");
        }
    }

    private static int parseInt(String value, String field) {
        try {
            return Integer.parseInt(value);
        } catch (NumberFormatException error) {
            throw new IllegalArgumentException(field + " is not an integer");
        }
    }

    private static int parseIntRange(String value, String field, int min, int max) {
        int parsed = parseInt(value, field);
        if (parsed < min || parsed > max) {
            throw new IllegalArgumentException(field + " outside " + min + "..." + max);
        }
        return parsed;
    }

    private static double finiteDouble(String value, String field) {
        try {
            double parsed = Double.parseDouble(value);
            if (!Double.isFinite(parsed)) {
                throw new IllegalArgumentException(field + " is not finite");
            }
            return parsed;
        } catch (NumberFormatException error) {
            throw new IllegalArgumentException(field + " is not a number");
        }
    }

    private static List<PacketFieldAnnotation> annotations(String[] values) {
        String[] labels = {
                "Type", "Firmware millis", "Sequence", "Weight", "Flow",
                "Status", "Firmware quality", "Battery", "HX711 cadence", "USB dropped"
        };
        PacketFieldSemantic[] semantics = {
                PacketFieldSemantic.HEADER, PacketFieldSemantic.TIMESTAMP, PacketFieldSemantic.SEQUENCE,
                PacketFieldSemantic.WEIGHT, PacketFieldSemantic.FLOW, PacketFieldSemantic.STATUS,
                PacketFieldSemantic.QUALITY, PacketFieldSemantic.BATTERY, PacketFieldSemantic.SAMPLE_RATE,
                PacketFieldSemantic.PAYLOAD
        };
        List<PacketFieldAnnotation> result = new ArrayList<>();
        int offset = 0;
        for (int i = 0; i < values.length; i++) {
            int length = values[i].getBytes(StandardCharsets.UTF_8).length;
            result.add(PacketFieldAnnotation.of(offset, offset + length, labels[i], values[i], semantics[i]));
            offset += length + 1;
        }
        return result;
    }

	    static final class ParseResult {
        final boolean ignored;
        final ScaleSample sample;
        final RawScalePacket packet;
        final String rejectionReason;

        private ParseResult(boolean ignored, ScaleSample sample, RawScalePacket packet, String rejectionReason) {
            this.ignored = ignored;
            this.sample = sample;
            this.packet = packet;
            this.rejectionReason = rejectionReason;
        }

        static ParseResult ignored() {
            return new ParseResult(true, null, null, null);
        }

        static ParseResult sample(ScaleSample sample, RawScalePacket packet) {
            return new ParseResult(false, sample, packet, null);
        }

        static ParseResult rejected(String reason) {
            return new ParseResult(false, null, null, reason);
	    }
	}
}
